### PR TITLE
Improve sender checks

### LINKED ISSUE #390: Can't send from an email account that has forwarding
I've setup an email account (for example  "myname@example.com").

I then turn on forwarding so that all emails to "myname@example.com" will be forwarded to "mygmail@gmail.com".

However when doing this I can no longer send emails from "myname@example.com", when doing (either via SMTP directly or roundcube) I get error such as:

> An error occurred while sending mail. The mail server responded:  
> 5.7.1 <myname@example.com>: Sender address rejected: not owned by user myname@example.com.
>  Please check the message recipient "someoneelse@yahoo.com" and try again.

### PR DESCRIPTION
**Pull request by kaiyou**
This fixes #390 and cleans most of the postfix views on top of the database abstraction code

### CODE REVIEW

**Comment by muhlemmer** [comment]
Please base the merge against master.

**Comment by kaiyou** [comment]
This should be fixed now

**Comment by muhlemmer** [comment]
I've tried to merge this on my test server, after handling the conflict I have all sorts of errors on the models. So probably I didn't do it right.
Could you be so friendly to handle the conflict and see if it works on your side?

I've been flipping though all kind of PR's the last hours. Different password encryption schemes and everything. I think I messed up somewhere. I'm giving up for today.

**Comment by kaiyou** [comment]
Just fixed the merge conflict. I will have to test again, there might be some issues with interfering changes.

**Comment by muhlemmer** [comment]
After rebuild, I'm still seeing the model errors. Some info:
- IMAP login works
- SMTP login errors

Logs:
```
front_1      | 2018/10/20 22:43:14 [info] 10#10: *23 client 5.15.255.211:46152 connected to 0.0.0.0:587
front_1      | 127.0.0.1 - - [20/Oct/2018:22:43:15 +0000] "GET /auth/email HTTP/1.0" 200 0 "-" "-"
admin_1      | 172.28.0.4 - - [20/Oct/2018:22:43:15 +0000] "GET /internal/auth/email HTTP/1.0" 200 - "-" "-"
smtp_1       | postfix/smtpd[114]: connect from mailu_front_1.mailu_default[172.28.0.4]
antispam_1   | 2018-10-20 22:43:15 #7(rspamd_proxy) <a49089>; proxy; proxy_accept_socket: accepted milter connection from 172.28.0.8 port 59984
antispam_1   | 2018-10-20 22:43:15 #7(rspamd_proxy) <a49089>; milter; rspamd_milter_process_command: got connection from 172.28.0.4:56364
antispam_1   | 2018-10-20 22:43:15 #7(rspamd_proxy) <a49089>; proxy; proxy_milter_finish_handler: finished milter connection
front_1      | 2018/10/20 22:43:15 [info] 10#10: *23 client logged in, client: 5.15.255.211 using starttls, server: 0.0.0.0:587, login: "admin@usrpro.io", upstream: 172.28.0.8:10025
antispam_1   | 2018-10-20 22:43:15 #7(rspamd_proxy) <4282ab>; proxy; proxy_accept_socket: accepted milter connection from 172.28.0.8 port 59986
admin_1      | [2018-10-20 22:43:15,617] ERROR in app: Exception on /internal/postfix/transport/* [GET]
admin_1      | Traceback (most recent call last):
admin_1      |   File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1982, in wsgi_app
admin_1      |     response = self.full_dispatch_request()
admin_1      |   File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1614, in full_dispatch_request
admin_1      |     rv = self.handle_user_exception(e)
admin_1      |   File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1517, in handle_user_exception
admin_1      |     reraise(exc_type, exc_value, tb)
admin_1      |   File "/usr/local/lib/python3.7/site-packages/flask/_compat.py", line 33, in reraise
admin_1      |     raise value
admin_1      |   File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1612, in full_dispatch_request
admin_1      |     rv = self.dispatch_request()
admin_1      |   File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1598, in dispatch_request
admin_1      |     return self.view_functions[rule.endpoint](**req.view_args)
admin_1      |   File "/app/mailu/internal/views/postfix.py", line 30, in postfix_transport
admin_1      |     localpart, domain = models.Email.resolve_domain(email)
admin_1      |   File "/app/mailu/models.py", line 227, in resolve_domain
admin_1      |     alternative = models.Alternative.query.get(domain_name)
admin_1      | NameError: name 'models' is not defined
admin_1      | 172.28.0.8 - - [20/Oct/2018:22:43:15 +0000] "GET /internal/postfix/transport/* HTTP/1.1" 500 291 "-" "Python/3.6 aiohttp/3.4.4"
smtp_1       | root (ERROR): Error when processing request
smtp_1       | Traceback (most recent call last):
smtp_1       |   File "/usr/lib/python3.6/site-packages/podop/postfix.py", line 105, in process_request
smtp_1       |     result = await table.get(key)
admin_1      | [2018-10-20 22:43:15,631] ERROR in app: Exception on /internal/postfix/transport/* [GET]
admin_1      | Traceback (most recent call last):
admin_1      |   File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1982, in wsgi_app
admin_1      |     response = self.full_dispatch_request()
admin_1      |   File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1614, in full_dispatch_request
admin_1      |     rv = self.handle_user_exception(e)
admin_1      |   File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1517, in handle_user_exception
admin_1      |     reraise(exc_type, exc_value, tb)
admin_1      |   File "/usr/local/lib/python3.7/site-packages/flask/_compat.py", line 33, in reraise
admin_1      |     raise value
admin_1      |   File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1612, in full_dispatch_request
admin_1      |     rv = self.dispatch_request()
admin_1      |   File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1598, in dispatch_request
admin_1      |     return self.view_functions[rule.endpoint](**req.view_args)
admin_1      |   File "/app/mailu/internal/views/postfix.py", line 30, in postfix_transport
admin_1      |     localpart, domain = models.Email.resolve_domain(email)
admin_1      |   File "/app/mailu/models.py", line 227, in resolve_domain
admin_1      |     alternative = models.Alternative.query.get(domain_name)
admin_1      | NameError: name 'models' is not defined
admin_1      | 172.28.0.8 - - [20/Oct/2018:22:43:15 +0000] "GET /internal/postfix/transport/* HTTP/1.1" 500 291 "-" "Python/3.6 aiohttp/3.4.4"
smtp_1       |   File "/usr/lib/python3.6/site-packages/podop/table.py", line 34, in get
smtp_1       |     raise Exception(request.status)
smtp_1       | postfix/trivial-rewrite[117]: warning: socketmap:unix:/tmp/podop.socket:transport socketmap server temporary error: unknown error
smtp_1       | Exception: 500
smtp_1       | root (ERROR): Error when processing request
smtp_1       | Traceback (most recent call last):
smtp_1       |   File "/usr/lib/python3.6/site-packages/podop/postfix.py", line 105, in process_request
smtp_1       |     result = await table.get(key)
smtp_1       |   File "/usr/lib/python3.6/site-packages/podop/table.py", line 34, in get
smtp_1       |     raise Exception(request.status)
smtp_1       | Exception: 500
smtp_1       | postfix/trivial-rewrite[117]: warning: socketmap:unix:/tmp/podop.socket:transport lookup error for "*"
smtp_1       | postfix/trivial-rewrite[117]: warning: socketmap:unix:/tmp/podop.socket:transport socketmap server temporary error: unknown error
smtp_1       | postfix/trivial-rewrite[117]: warning: socketmap:unix:/tmp/podop.socket:transport lookup error for "*"
```

**Comment by muhlemmer** [comment]
I've managed to take out the python errors on "models". However, now I'm getting temporary lookup failures. Admin logs are clean around the time time frame. Just some logs from smtp and front.
```
smtp_1       | 2018-10-23T08:49:19.975330882Z postfix/smtpd[792]: NOQUEUE: reject: RCPT from unknown[188.27.138.94]: 451 4.3.0 <admin@usrpro.io>: Temporary lookup failure; from=<admin@usrpro.io> to=<tim@usrpro.com> proto=ESMTP helo=<[192.168.100.5]>
front_1      | 2018-10-23T08:49:19.090998761Z 2018/10/23 08:49:19 [info] 10#10: *519 client 188.27.138.94:38826 connected to 0.0.0.0:587
front_1      | 2018-10-23T08:49:19.101433766Z 2018/10/23 08:49:19 [error] 10#10: *519 "188-27-138-94.rdsnet.ro" could not be resolved (3: Host not found) while in resolving client hostname, client: 188.27.138.94, server: 0.0.0.0:587
front_1      | 2018-10-23T08:49:19.841800378Z 127.0.0.1 - - [23/Oct/2018:08:49:19 +0000] "GET /auth/email HTTP/1.0" 200 0 "-" "-"
front_1      | 2018-10-23T08:49:19.845991496Z 2018/10/23 08:49:19 [info] 10#10: *519 client logged in, client: 188.27.138.94 using starttls, server: 0.0.0.0:587, login: "admin@usrpro.io", upstream: 172.29.0.8:10025
```

File: core/admin/mailu/internal/views/postfix.py [hunk 1]
```diff
@@ -18,37 +18,32 @@ def postfix_mailbox_map(email):
...
+        return flask.jsonify(domain_name)
+    destination = models.Email.resolve_destination(localpart, domain_name)
+    return flask.jsonify(",".join(destination)) if destination else flask.abort(404)
 
 
 @internal.route("/postfix/transport/<email>")
 def postfix_transport(email):
-    localpart, domain = email.split('@', 1) if '@' in email else (None, email)
-    relay = models.Relay.query.get(domain) or flask.abort(404)
+    localpart, domain_name = models.Email.resolve_domain(email)
```
Comment (muhlemmer) [inline core/admin/mailu/internal/views/postfix.py]: At one point it seems and asterisk (`*`) is passed here, which fails against the IDNA type.

```
admin_1      | The above exception was the direct cause of the following exception:
admin_1      | 
admin_1      | Traceback (most recent call last):
admin_1      |   File "/usr/lib/python3.6/site-packages/flask/app.py", line 1982, in wsgi_app
admin_1      |     response = self.full_dispatch_request()
admin_1      |   File "/usr/lib/python3.6/site-packages/flask/app.py", line 1614, in full_dispatch_request
admin_1      |     rv = self.handle_user_exception(e)
admin_1      |   File "/usr/lib/python3.6/site-packages/flask/app.py", line 1517, in handle_user_exception
admin_1      |     reraise(exc_type, exc_value, tb)
admin_1      |   File "/usr/lib/python3.6/site-packages/flask/_compat.py", line 33, in reraise
admin_1      |     raise value
admin_1      |   File "/usr/lib/python3.6/site-packages/flask/app.py", line 1612, in full_dispatch_request
admin_1      |     rv = self.dispatch_request()
admin_1      |   File "/usr/lib/python3.6/site-packages/flask/app.py", line 1598, in dispatch_request
admin_1      |     return self.view_functions[rule.endpoint](**req.view_args)
admin_1      |   File "/app/mailu/internal/views/postfix.py", line 30, in postfix_transport
admin_1      |     localpart, domain_name = models.Email.resolve_domain(email)
admin_1      |   File "/app/mailu/models.py", line 231, in resolve_domain
admin_1      |     alternative = Alternative.query.get(domain_name)
admin_1      |   File "/usr/lib/python3.6/site-packages/sqlalchemy/orm/query.py", line 882, in get
admin_1      |     ident, loading.load_on_ident)
admin_1      |   File "/usr/lib/python3.6/site-packages/sqlalchemy/orm/query.py", line 916, in _get_impl
admin_1      |     return fallback_fn(self, key)
admin_1      |   File "/usr/lib/python3.6/site-packages/sqlalchemy/orm/loading.py", line 232, in load_on_ident
admin_1      |     return q.one()
admin_1      |   File "/usr/lib/python3.6/site-packages/sqlalchemy/orm/query.py", line 2848, in one
admin_1      |     ret = self.one_or_none()
admin_1      |   File "/usr/lib/python3.6/site-packages/sqlalchemy/orm/query.py", line 2818, in one_or_none
admin_1      |     ret = list(self)
admin_1      |   File "/usr/lib/python3.6/site-packages/sqlalchemy/orm/query.py", line 2889, in __iter__
admin_1      |     return self._execute_and_instances(context)
admin_1      |   File "/usr/lib/python3.6/site-packages/sqlalchemy/orm/query.py", line 2912, in _execute_and_instances
admin_1      |     result = conn.execute(querycontext.statement, self._params)
admin_1      |   File "/usr/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 948, in execute
admin_1      |     return meth(self, multiparams, params)
admin_1      |   File "/usr/lib/python3.6/site-packages/sqlalchemy/sql/elements.py", line 269, in _execute_on_connection
admin_1      |     return connection._execute_clauseelement(self, multiparams, params)
admin_1      |   File "/usr/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1060, in _execute_clauseelement
admin_1      |     compiled_sql, distilled_params
admin_1      |   File "/usr/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1132, in _execute_context
admin_1      |     None, None)
admin_1      |   File "/usr/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1413, in _handle_dbapi_exception
admin_1      |     exc_info
admin_1      |   File "/usr/lib/python3.6/site-packages/sqlalchemy/util/compat.py", line 203, in raise_from_cause
admin_1      |     reraise(type(exception), exception, tb=exc_tb, cause=cause)
admin_1      |   File "/usr/lib/python3.6/site-packages/sqlalchemy/util/compat.py", line 186, in reraise
admin_1      |     raise value.with_traceback(tb)
admin_1      |   File "/usr/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1127, in _execute_context
admin_1      |     context = constructor(dialect, self, conn, *args)
admin_1      |   File "/usr/lib/python3.6/site-packages/sqlalchemy/engine/default.py", line 667, in _init_compiled
admin_1      |     param.append(processors[key](compiled_params[key]))
admin_1      |   File "/usr/lib/python3.6/site-packages/sqlalchemy/sql/type_api.py", line 1160, in process
admin_1      |     return process_param(value, dialect)
admin_1      |   File "/app/mailu/models.py", line 25, in process_bind_param
admin_1      |     return idna.encode(value).decode("ascii")
admin_1      |   File "/usr/lib/python3.6/site-packages/idna/core.py", line 355, in encode
admin_1      |     result.append(alabel(label))
admin_1      |   File "/usr/lib/python3.6/site-packages/idna/core.py", line 265, in alabel
admin_1      |     raise IDNAError('The label {0} is not a valid A-label'.format(label))
admin_1      | sqlalchemy.exc.StatementError: (idna.core.IDNAError) The label b'*' is not a valid A-label [SQL: 'SELECT alternative.created_at AS alternative_created_at, alternative.updated_at AS alternative_updated_at, alternative.comment AS alternative_comment, alternative.name AS alternative_name, alternative.domain_name AS alternative_domain_name \nFROM alternative \nWHERE alternative.name = ?'] [parameters: [{'%(140545964121000 param)s': '*'}]]
```

File: core/admin/mailu/models.py [hunk 2]
```diff
@@ -221,6 +221,28 @@ def sendmail(self, subject, body):
             msg['To'] = to_address
             smtp.sendmail(from_address, [to_address], msg.as_string())
 
+    @classmethod
+    def resolve_domain(cls, email):
+        localpart, domain_name = email.split('@', 1) if '@' in email else (None, email)
+        alternative = Alternative.query.get(domain_name)
```
Comment (muhlemmer) [inline core/admin/mailu/models.py]: I've inserted `print(domain_name, file=sys.stderr,flush=True)` before this line. This printed the following (between exceptions):
```
admin_1      | 192.168.203.2 - - [04/Dec/2018:10:36:41 +0000] "GET /internal/postfix/transport/* HTTP/1.1" 500 291 "-" "Python/3.6 aiohttp/3.4.4"
admin_1      | 192.168.203.2 - - [04/Dec/2018:10:36:41 +0000] "GET /internal/postfix/domain/usrpro.io HTTP/1.1" 200 12 "-" "Python/3.6 aiohttp/3.4.4"
admin_1      | usrpro.io
admin_1      | 192.168.203.2 - - [04/Dec/2018:10:36:41 +0000] "GET /internal/postfix/transport/usrpro.io HTTP/1.1" 404 233 "-" "Python/3.6 aiohttp/3.4.4"
admin_1      | usrpro.io
admin_1      | 192.168.203.2 - - [04/Dec/2018:10:36:41 +0000] "GET /internal/postfix/transport/admin@usrpro.io HTTP/1.1" 404 233 "-" "Python/3.6 aiohttp/3.4.4"
admin_1      | *
```

File: core/admin/mailu/models.py [hunk 3]
```diff
@@ -245,7 +267,7 @@ class User(Base, Email):
 
     # Filters
     forward_enabled = db.Column(db.Boolean(), nullable=False, default=False)
-    forward_destination = db.Column(db.String(255), nullable=True, default=None)
+    forward_destination = db.Column(CommaSeparatedList(), nullable=True, default=None)
```
Comment (muhlemmer) [inline core/admin/mailu/models.py]: Reverting this back to db.String prevents the following error. It does not seem to impact the other reported errors.

```
$ docker-compose exec admin python3 manage.py admin root usrpro.io password
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1127, in _execute_context
    context = constructor(dialect, self, conn, *args)
  File "/usr/lib/python3.6/site-packages/sqlalchemy/engine/default.py", line 667, in _init_compiled
    param.append(processors[key](compiled_params[key]))
  File "/usr/lib/python3.6/site-packages/sqlalchemy/sql/type_api.py", line 1160, in process
    return process_param(value, dialect)
  File "/app/mailu/models.py", line 63, in process_bind_param
    raise TypeError("Shoud be a list")
TypeError: Shoud be a list
```

File: core/admin/mailu/internal/views/postfix.py (hunk 1 above)
Comment (muhlemmer) [reply core/admin/mailu/internal/views/postfix.py]: So on master the asterisk was always there, just it seems never to reach the type-check.

```
192.168.203.2 - - [04/Dec/2018:13:07:22 +0000] "GET /internal/postfix/transport/* HTTP/1.1" 404 233 "-" "Python/3.6 aiohttp/3.4.4"
```
It seems to always result in a 404. So if we catch early in `postfix_transport(email)` and return a 404, it should be ok.

**Pull request merged**